### PR TITLE
feat(vfr-rt): Phase C — student exam UI (briefing, runner, results) (#697)

### PR DESCRIPTION
**Pull request by okpilot**
## Phase C — VFR RT Student UI (#697)

Frontend-only slice of the VFR RT Slovenia mock exam. **No migrations / no new RPCs** — consumes the already-merged Phase A/B RPCs (`start_vfr_rt_exam_session`, `get_vfr_rt_exam_questions` mig 105, `submit_vfr_rt_exam_answers`, `get_vfr_rt_exam_results` mig 106) and Phase B server actions.

> ⚠️ **Not for merge yet — manual eval requested.** This is the first renderable VFR-RT surface. The feature stays **dark** in prod (no `exam_configs` row ⇒ `exam_config_required`), so merging is safe but the UI should be eyeballed first. Leaving open for review.

### What's here (spec tasks C.1–C.5)
- **C.1 Briefing/landing** — `/app/vfr-rt-exam`: Server Component resumes an in-progress session (redirect) or renders the briefing + Start button (`startVfrRtExam` → navigate into the runner).
- **C.2 In-progress runner** — `/app/vfr-rt-exam/in-progress/[sessionId]`: timed runner with per-type renderers (short-answer / dialog-fill / MC), a localStorage answer buffer, a server-derived `ExamCountdownTimer`, part-progress, and a **synchronous double-submit guard** (manual Finish + timer-expiry race).
- **C.3 Per-type renderers** + **C.4 part-progress bar** — leaf presentational components; the dialog parser strips canonicals (`{{n}}` markers only).
- **C.5 Results/review** — `/app/vfr-rt-exam/results/[sessionId]`: per-part bars with a 75% threshold marker, overall pass/fail, and a per-question review. **Two-RPC merge** — display fields from 105, scores + revealed key + explanations from the `ended_at`-gated 106. Dedicated `vfr-rt-review-row` (not the shared MC report row, which can't show per-blank dialog review).

### Security
Correct-answer material is reachable **only** post-completion via the `ended_at`-gated 106 RPC; the in-flight path (105) is answer-stripped. All result components are Server Components (keys render server-side, never in client props). `quiz_sessions` reads scope `student_id = auth.uid()` explicitly (security.md §11). Pre-push security-auditor: APPROVED.

### Quality
~30 new files. Full web suite **4197 tests green**, types + biome clean. Each slice ran plan-critic / impl-critic + the full post-commit fleet; the branch passed a PR-level semantic sweep and 5 rounds of CodeRabbit-local triage (all real findings applied).

### Not in scope (tracked)
- Phase D (admin authoring), Phase E (Playwright lifecycle E2E `spec task E.1`, red-team #825).
- A `code-style.md` rule promotion for the async-submit re-entry guard (learner count=3) — to be filed separately so it doesn't expand this PR.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Delivered VFR RT mock exam student UI (Phase C): briefing, interactive question runner, and results breakdown with question review.
  * Added renderers for multiple-choice, short-answer, and dialog-fill, including hiding canonical answer content in dialog-fill.
  * Implemented browser-based answer persistence plus per-part progress and pass/fail indicators.
  * Added authenticated in-progress and results pages with automatic redirects.

* **Bug Fixes**
  * Improved finish submission reliability for timer/click races and allowed graceful empty-answer submissions on expiry.
  * Fixed vacuous-truth scoring so unanswered items no longer appear as passed.

* **Tests**
  * Expanded UI, parsing/payload, and server query test coverage for the full exam flow, including edge cases and localStorage failure scenarios.

### CODE REVIEW

File: apps/web/app/app/vfr-rt-exam/_components/vfr-rt-part-nav.tsx [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  return (
+    <div className="flex items-center justify-between gap-4">
+      <Button variant="outline" onClick={onPrev} disabled={currentIndex === 0}>
+        Previous
+      </Button>
+      <p className="text-sm text-muted-foreground">
+        Question {currentIndex + 1} of {total} · {partLabel}
+      </p>
+      <Button variant="outline" onClick={onNext} disabled={currentIndex === total - 1}>
+        Next
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/vfr-rt-exam/_components/vfr-rt-part-nav.tsx:37]: Fixed in 76ea7289 — VfrRtPartNav now disables both buttons when total<=1 and clamps the displayed index to 0, so a zero-question set no longer renders "Question 1 of 0" with Next enabled.

File: apps/web/app/app/vfr-rt-exam/_components/vfr-rt-part-nav.test.tsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    const { onNext } = renderNav({ currentIndex: 1 })
+    await userEvent.click(screen.getByRole('button', { name: /next/i }))
+    expect(onNext).toHaveBeenCalledOnce()
+  })
+
+  it('shows the current position and part label', () => {
+    renderNav({ currentIndex: 1, total: 3, partLabel: 'Part 2' })
+    expect(screen.getByText('Question 2 of 3 · Part 2')).toBeInTheDocument()
+  })
+})
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/vfr-rt-exam/_components/vfr-rt-part-nav.test.tsx:55]: Fixed in 76ea7289 — added a zero-total regression test asserting both buttons disabled and "Question 0 of 0".

File: apps/web/app/app/vfr-rt-exam/_hooks/use-vfr-rt-answers.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+function loadAnswers(sessionId: string): AnswersMap {
+  try {
+    const raw = localStorage.getItem(storageKey(sessionId))
+    if (!raw) return {}
+    const parsed: unknown = JSON.parse(raw)
+    // typeof [] === 'object', so guard arrays out before casting to the keyed map.
+    if (typeof parsed !== 'object' || parsed === null || Array.isArray(parsed)) return {}
+    return parsed as AnswersMap
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/vfr-rt-exam/_hooks/use-vfr-rt-answers.ts]: Fixed in 76ea7289 — loadAnswers now validates each entry's shape (mc/short must be strings; blanks keeps only string values; non-object entries dropped). Extracted to vfr-rt-answer-storage.ts (+ co-located test) to keep the hook under the 80-line cap.

File: apps/web/app/app/vfr-rt-exam/_utils/build-vfr-rt-payload.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    if (question.question_type === 'dialog_fill') {
+      for (const [blankIndex, text] of Object.entries(answer.blanks ?? {})) {
+        if (text.trim()) {
+          entries.push({
+            questionId: question.id,
+            blankIndex: Number(blankIndex),
+            responseText: text,
+          })
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/vfr-rt-exam/_utils/build-vfr-rt-payload.ts:39]: Fixed in 76ea7289 — buildVfrRtPayload now emits the trimmed value for short answers and dialog blanks, not the raw padded input.

File: apps/web/app/app/vfr-rt-exam/_utils/build-vfr-rt-payload.test.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    }
+    const result = buildVfrRtPayload(questions, answers)
+    expect(result).toHaveLength(2)
+    expect(result).toContainEqual({
+      questionId: 'q-dialog',
+      blankIndex: 0,
+      responseText: 'cleared',
+    })
+    expect(result).toContainEqual({ questionId: 'q-dialog', blankIndex: 2, responseText: 'final' })
+  })
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/vfr-rt-exam/_utils/build-vfr-rt-payload.test.ts:99]: Fixed in 76ea7289 — added a test asserting '  QNH  ' → 'QNH' and '  cleared  ' → 'cleared' for dialog blanks.

File: apps/web/app/app/vfr-rt-exam/actions/submit.test.ts [hunk 6]
```diff
@@ -81,11 +81,19 @@ describe('submitVfrRtExam — input validation', () => {
     mockGetUser.mockResolvedValue({ data: { user: { id: 'u1' } } })
   })
 
-  it('rejects an empty answers array', async () => {
+  it('returns success with the expired flag when submitting an empty answers array', async () => {
+    // A timer expiry with zero answers submits []. The action must NOT reject it
+    // client-side — the RPC's expiry branch grades+closes it and returns expired:true.
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/vfr-rt-exam/actions/submit.test.ts]: Fixed in 76ea7289 — removed the title-paraphrasing comment; the production rationale lives in submit.ts.

File: apps/web/lib/queries/vfr-rt-exam.ts [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+  const { data, error: rpcError } = await rpc<VfrRtQuestion[]>(
+    supabase,
+    'get_vfr_rt_exam_questions',
+    { p_session_id: sessionId },
+  )
+  if (rpcError || !data || data.length === 0) {
+    if (rpcError) console.error('[getVfrRtInProgress] RPC error:', rpcError.message)
+    else console.error('[getVfrRtInProgress] RPC returned no questions')
+    return { status: 'not_found' }
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/lib/queries/vfr-rt-exam.ts:135]: Fixed in 76ea7289 — getVfrRtInProgress now guards with !Array.isArray(data), so a truthy non-array payload returns not_found instead of slipping a malformed questions set into the active runner.

File: apps/web/lib/queries/vfr-rt-exam.test.ts [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+  })
+
+  it('returns not_found without querying the session when auth fails', async () => {
+    mockGetUser.mockResolvedValueOnce({ data: { user: null }, error: { message: 'expired' } })
+    const result = await getVfrRtInProgress('sess-1')
+    expect(result).toEqual({ status: 'not_found' })
+    expect(mockFrom).not.toHaveBeenCalled()
+    expect(mockRpc).not.toHaveBeenCalled()
+  })
+})
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/lib/queries/vfr-rt-exam.test.ts:238]: Fixed in 76ea7289 — added a case where the RPC resolves with a truthy non-array ({}) asserting not_found + the log.

File: apps/web/lib/queries/vfr-rt-results.ts [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    for (const q of questionsData) {
+      displayMap.set(q.id, q)
+    }
+  }
+
+  const rows: VfrRtReviewRow[] = resultsData.questions.map((q) => {
+    const display = displayMap.get(q.question_id)
+    // every() is vacuously true on []; an unanswered question (timer-expiry or
+    // partial submit returns answers: []) must show incorrect, not a ✓.
+    const isCorrect = q.answers.length > 0 && q.answers.every((a) => a.is_correct)
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/lib/queries/vfr-rt-results.ts:125]: Fixed in 76ea7289 — getVfrRtResults now uses Array.isArray(questionsData) before the for...of (logs + degrades on non-array), and throws if a question's answers field is not an array.

File: apps/web/lib/queries/vfr-rt-results.test.ts [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+      .mockResolvedValueOnce({ data: sampleResultsJson, error: null })
+      .mockResolvedValueOnce({ data: null, error: { message: 'rpc failed' } })
+    const consoleSpy = vi.spyOn(console, 'error').mockImplementation(() => {})
+    const result = await getVfrRtResults('sess-1')
+    expect(result).not.toBeNull()
+    expect(result!.rows[0]!.options).toBeNull()
+    expect(result!.rows[0]!.questionImageUrl).toBeNull()
+    expect(consoleSpy).toHaveBeenCalledWith('[getVfrRtResults] Questions RPC error:', 'rpc failed')
+    consoleSpy.mockRestore()
+  })
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/lib/queries/vfr-rt-results.test.ts:182]: Fixed in 76ea7289 — added a test where the display RPC returns a non-array payload, asserting rows still return with null options/image + the degradation log.

File: apps/web/app/app/vfr-rt-exam/_hooks/vfr-rt-answer-storage.ts [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    for (const [qId, v] of Object.entries(parsed)) {
+      if (!isPlainObject(v)) continue
+      const next: AnswerState = {}
+      if (typeof v.mc === 'string') next.mc = v.mc
+      if (typeof v.short === 'string') next.short = v.short
+      if (isPlainObject(v.blanks)) {
+        const b: Record<number, string> = {}
+        for (const [k, val] of Object.entries(v.blanks)) {
+          if (typeof val === 'string') b[Number(k)] = val
+        }
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/vfr-rt-exam/_hooks/vfr-rt-answer-storage.ts:39]: Fixed in ac74b315 — loadAnswers now gates blank keys on Number.isInteger(idx) && idx >= 0, dropping NaN/negative/decimal keys before they reach the payload. Added a test covering "foo"/"-1"/"1.5".
